<a href="https://colab.research.google.com/github/AIVIETNAM-AIO-HUYTRUONG/Artificial-Intelligence/blob/basic/02.%20Essential%20NumPy%20and%20Pandas/Chapter%202%3A%20Getting%20Started%20with%20NumPy%20Arrays/03_built_in_array_creation_functions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Built-in Array Creation Functions**

- Việc tạo array bằng các hàm chuyên biệt của NumPy thường thuận tiện và hiệu quả hơn. Các hàm này cho phép bạn tạo array với cấu trúc hoặc giá trị khởi tạo cụ thể mà không cần tạo Python list trước. NumPy cung cấp một số hàm thường được sử dụng.

## **1. Creating Sequences with `arange`**

- Tương tự hàm tích hợp sẵn `range` của Python, hàm `arange` của NumPy tạo một array chứa một chuỗi các giá trị cách đều trong một khoảng cho trước. Tuy nhiên, không giống với `range` vốn tạo một **generator**, `arange` trả về trực tiếp một NumPy array.

- Cú pháp cơ bản là `np.arange(start, stop, step)`, trong đó:
  - `start`: Giá trị bắt đầu của khoảng, được bao gồm trong kết quả. Mặc định là 0 nếu không được cung cấp.
  - `stop`: Giá trị kết thúc của khoảng, không được bao gồm trong kết quả.
  - `step`: Khoảng cách giữa các giá trị. Mặc định là `1`.

In [ ]:
import numpy as np

# Create an array from 0 up to (but not including) 5
arr1 = np.arange(5)
print(arr1)
# Output: [0 1 2 3 4]

# Create an array from 2 up to (but not including) 8
arr2 = np.arange(2, 8)
print(arr2)
# Output: [2 3 4 5 6 7]

# Create an array from 1 to 10 with a step of 2
arr3 = np.arange(1, 10, 2)
print(arr3)
# Output: [1 3 5 7 9]

[0 1 2 3 4]
[2 3 4 5 6 7]
[1 3 5 7 9]


- Lưu ý rằng `arange`, giống như `range` của Python, không bao gồm giá trị `stop` trong kết quả. Ngoài ra, `arange` có thể sử dụng `step` dạng **floating-point**, nhưng cần thận trọng vì có thể xuát hiện **sai số dấu phẩy động (floating-point inaccuracies)**. Đối với `step` không phải số nguyên, khi số lượng điểm chính xác quan trọng hơn, `linspace` thường được ưu tiên sử dụng.


###  **1.1. Vấn đề với `np.arange` và Floating-point Step**:

- Khi sử dụng `np.arange()` với **bước nhảy số thực (floating-point step)**, kết quả đôi khi có thể bị ảnh hưởng bởi sai số biểu diễn số thực trong máy tính.

In [ ]:
import numpy as np

arr = np.arange(0, 0.4, 0.1)

print(arr)
print(arr[3] == 0.3)
print(repr(arr[3]))

[0.  0.1 0.2 0.3]
False
np.float64(0.30000000000000004)


- Mặc dù NumPy hiển thị phần tử cuối là `0.3`, giá trị thực tế được lưu có thể gần với: `0.30000000000000004`

- Điều này xảy ra vì một số giá trị thập phân như `0.1` không thể được biểu diễn chính xác dưới dạng số nhị phân.

- Vì vậy, khi cần tạo các giá trị cách đều bằng số thực, đặc biệt khi muốn kiểm soát chính xác **số lượng phần tử**, `np.linspace()` thường là lựa chọn phù hợp hơn.

### **1.2. Sự ưu việt của `np.linspace`**

- `np.linspace(start, stop, num)` tạo ra `num` giá trị cách đều từ `start` đến `stop`, trong đó `stop` mặc định được bao gồm.

In [ ]:
import numpy as np

arr = np.linspace(0, 0.3, 4)

print(arr)

[0.  0.1 0.2 0.3]


- `np.linspace(0, 0.3, 4)` tạo ra chính xác `4` giá trị cách đều từ `0` đến `0.3`.

## **2. Creating Arrays of Zeros or Ones**

- Thông thường, bạn cần khởi tạo một array có kích thước cụ thể với các giá trị giữ chỗ, thường là số `0` hoặc số `1`. NumPy cung cấp `zeros` và `ones` cho mục đích này.
  - `np.zeros(shape, dtype=float)`: Tạo một array được lấp đầy bằng số `0`.
  - `np.ones(shape, dtype=float)`: Tạo một array được lấp đầy bằng số `1`.

- Đối số `shape` dùng để khai báo **kích thước của mảng** dưới dạng các con số trong dấu ngoặc tròn `()`.

  - Khi viết `(3,)`: Tạo mảng `1` chiều gồm `3` phần tử.

  - Khi viết `(2, 4)`: Tạo mảng `2` chiều gồm `2` hàng và `4` cột.

- `dtype` dùng để chọn kiểu dữ liệu cho các phần tử (như số nguyên hay số thực). Nếu bạn không ghi gì, NumPy sẽ tự động để mặc định là số thực (float64).

In [ ]:
# Create a 1D array of 4 zeros (default dtype is float)
zeros_arr_1d = np.zeros(4)
print(zeros_arr_1d)
# Output: [0. 0. 0. 0.]

# Create a 2x3 array of ones with integer type
ones_arr_2d_int = np.ones((2, 3), dtype=np.int64)
print(ones_arr_2d_int)
# Output:
# [[1 1 1]
#  [1 1 1]]

# Check the data type
print(ones_arr_2d_int.dtype)
# Output: int64

[0. 0. 0. 0.]
[[1 1 1]
 [1 1 1]]
int64


## **3. Creating Evenly Spaced Arrays with `linspace`**

- Đôi khi, bạn cần một array chứa một số lượng điểm cụ thể được phân bố đều giữa giá trị bắt đầu và giá trị kết thúc. Đây là trường hợp `linspace` hữu ích. Cú pháp là `np.linspace(start, stop, num=50)`, trong đó:
  - `start`: Giá trị bắt đầu của sequence, được bao gồm.
  - `stop`: Giá trị kết thúc của sequence, mặc định được bao gồm.
  - `num`: Số lượng mẫu cần tạo. Mặc định là 50.
- Không giống `arange`, `linspace` bao gồm giá trị `stop` trong array.

In [ ]:
# Create an array with 5 evenly spaced values between 0 and 1 (inclusive)
lin_arr1 = np.linspace(0, 1, 5)
print(lin_arr1)
# Output: [0.   0.25 0.5  0.75 1.  ]

# Create an array with 11 evenly spaced values between 0 and 10
lin_arr2 = np.linspace(0, 10, 11)
print(lin_arr2)
# Output: [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10.]

# You can exclude the endpoint if needed
lin_arr3 = np.linspace(0, 1, 5, endpoint=False)
print(lin_arr3)
# Output: [0.  0.2 0.4 0.6 0.8]

[0.   0.25 0.5  0.75 1.  ]
[ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10.]
[0.  0.2 0.4 0.6 0.8]


- `linspace` đặc biệt hữu ích để tạo tọa độ cho việc vẽ biểu đồ hoặc các mô phỏng.

## **4. Creating Identity Matrices with `eye`**

- **Ma trận đơn vị (identity matrix)** là một ma trận vuông, trong đó số col bằng số row, có các giá trị `1` trên một đường chéo từ góc trên bên phải xuống bên dưới và các vị trí còn lại đều bằng `0`.

- Hàm `eye` của NumPy được sử dụng để tạo ra các ma trận này. Cú pháp là `np.eye(N, dtype=float)`, trong đó `N` là số col và cũng là số row

In [ ]:
# Create a 3x3 identity matrix
identity_matrix = np.eye(3)
print(identity_matrix)
# Output:
# [[1. 0. 0.]
#  [0. 1. 0.]
#  [0. 0. 1.]]

# Create a 4x4 identity matrix with integer type
identity_matrix_int = np.eye(4, dtype=int)
print(identity_matrix_int)
# Output:
# [[1 0 0 0]
#  [0 1 0 0]
#  [0 0 1 0]
#  [0 0 0 1]]

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
[[1 0 0 0]
 [0 1 0 0]
 [0 0 1 0]
 [0 0 0 1]]


- Identity matrix là một thành phần cơ bản trong các phép toán **đại số tuyến tính (linear algebra)**.

## **5. Creating Arrays Filled with a Specific Value using `full`**

- `np.full` là hàm dùng để tạo một mảng có kích thước tùy chọn và lấp đầy tất cả các phần tử bằng **MỘT GIÁ TRỊ BẤT KỲ** do bạn chỉ định.

- Cú pháp là `np.full(shape, fill_value, dtype=None)`.

  - `fill_value` có thể là bất kỳ giá trị nào: từ số `0`, `1`, các số nguyên khác `(7, -5)`, số thập phân `(3.14)`, cho đến chuỗi chữ `("N/A")`



In [ ]:
import numpy as np

# Cả 2 lệnh dưới đây tạo ra kết quả HOÀN TOÀN GIỐNG NHAU:
a = np.ones((2, 3))       # Ngắn gọn hơn khi tạo số 1(float)
b = np.full((2, 3), 1)    # Kết quả y hệt!(int)

# Nhưng khi cần số 7 hoặc chữ "A", np.full là cách viết sạch đẹp nhất:
c = np.full((2, 3), 7)    # Mảng 2x3 toàn số 7
d = np.full((2, 3), "A")  # Mảng 2x3 toàn chữ "A"

print(a)
print(b)
print(c)
print(d)

[[1. 1. 1.]
 [1. 1. 1.]]
[[1 1 1]
 [1 1 1]]
[[7 7 7]
 [7 7 7]]
[['A' 'A' 'A']
 ['A' 'A' 'A']]


## **6. Generating Random Arrays**
- NumPy cũng bao gồm một **mô-đun con (submodule)** mạnh mẽ là `numpy.random`, dùng để tạo array chứa các số ngẫu nhiên được lấy từ nhiều distribution khác nhau. Dưới đây là một số ví dụ phổ biến:

  - `np.random.rand(d0, d1, ..., dn)`: Tạo một array có `shape` đã cho với các mẫu ngẫu nhiên từ **phân phối đều (uniform distribution)** trên khoảng `[0, 1)`.
  - `np.random.randn(d0, d1, ..., dn)`: Tạo một array có `shape` đã cho với các mẫu ngẫu nhiên từ **phân phối chuẩn tắc (standard normal distribution**) có mean bằng 0 và variance bằng 1.

  - `np.random.randint(low, high=None, size=None, dtype=int)`: Tạo một array có kích thước xác định với các số nguyên ngẫu nhiên từ `low` được bao gồm đến `high` không được bao gồm.

In [ ]:
# Create a 2x3 array with random values between 0 and 1
rand_arr = np.random.rand(2, 3)
print(rand_arr)
# Example Output (will vary):
# [[0.11150118 0.38348479 0.45066311]
#  [0.86726997 0.13023643 0.80802871]]

# Create a 1D array of size 4 with samples from standard normal distribution
randn_arr = np.random.randn(4)
print(randn_arr)
# Example Output (will vary):
# [-1.04782338  0.88233694 -0.22512731  0.280441  ]

# Create a 1D array of 5 random integers between 10 (inclusive) and 20 (exclusive)
randint_arr = np.random.randint(10, 20, size=5)
print(randint_arr)
# Example Output (will vary):
# [15 11 18 10 13]

[[0.78056317 0.7444296  0.59466437]
 [0.14862909 0.45319055 0.87474635]]
[-0.40581067  0.5975296  -1.71887446  0.80877534]
[12 12 14 19 14]


- Các hàm tạo số ngẫu nhiên này rất quan trọng đối với **mô phỏng (simulations)**, **mô hình hóa thống kê (statistical modeling)** và việc khởi tạo tham số trong các **thuật toán học máy (machine learning algorithms)**.

- Các hàm tích hợp này cung cấp những cách linh hoạt và hiệu quả để tạo NumPy array cho nhiều tác vụ tính toán khác nhau, tạo nền tảng cho nhiều **quy trình tính toán số (numerical workflows)** trong Python.